api/src/main/java/org/pmiops/workbench/api/WorkspacesController.java

```diff
@@ -171,31 +178,52 @@ private FirecloudWorkspaceId generateFirecloudWorkspaceId(String namespace, Stri
     } else if (workspace.getName().length() > 80) {
       throw new BadRequestException("Workspace name must be 80 characters or less");
     }
+
+    final boolean useBillingProjectBuffer = workbenchConfig.get().featureFlags.useBillingProjectBuffer;
+    if (useBillingProjectBuffer) {
+      if (Strings.isNullOrEmpty(workspace.getNamespace())) {
+        throw new BadRequestException("missing required field 'namespace'");
+      }
+      org.pmiops.workbench.db.model.Workspace existingWorkspace = workspaceService.getByName(
+              workspace.getNamespace(), workspace.getName());
+      if (existingWorkspace != null) {
+        throw new ConflictException(String.format(
+                "Workspace %s/%s already exists",
+                workspace.getNamespace(), workspace.getName()));
+      }
+    }
+
     User user = userProvider.get();
-    org.pmiops.workbench.db.model.Workspace existingWorkspace = workspaceService.getByName(
-        workspace.getNamespace(), workspace.getName());
-    if (existingWorkspace != null) {
-      throw new ConflictException(String.format(
-          "Workspace %s/%s already exists",
-          workspace.getNamespace(), workspace.getName()));
+    String workspaceNamespace;
+    if (useBillingProjectBuffer) {
+      BillingProjectBufferEntry bufferedBillingProject = billingProjectBufferService.assignBillingProject(user);
+      workspaceNamespace = bufferedBillingProject.getFireCloudProjectName();
+    } else {
+      workspaceNamespace = workspace.getNamespace();
     }
 
     // Note: please keep any initialization logic here in sync with CloneWorkspace().
-    FirecloudWorkspaceId workspaceId = generateFirecloudWorkspaceId(workspace.getNamespace(),
+    FirecloudWorkspaceId workspaceId = generateFirecloudWorkspaceId(workspaceNamespace,
         workspace.getName());
     FirecloudWorkspaceId fcWorkspaceId = workspaceId;
-    org.pmiops.workbench.firecloud.model.Workspace fcWorkspace = null;
-    for (int attemptValue = 0; attemptValue < MAX_FC_CREATION_ATTEMPT_VALUES; attemptValue++) {
-      try {
-        fcWorkspace = attemptFirecloudWorkspaceCreation(fcWorkspaceId);
-        break;
-      } catch (ConflictException e) {
-        if (attemptValue >= 5) {
-          throw e;
-        } else {
-          fcWorkspaceId =
-              new FirecloudWorkspaceId(workspaceId.getWorkspaceNamespace(),
-                  workspaceId.getWorkspaceName() + Integer.toString(attemptValue));
+    org.pmiops.workbench.firecloud.model.Workspace fcWorkspace;
+
+    if (useBillingProjectBuffer) {
```

Comment:
Why fork here? I think the existing retry logic is sketchy, but we probably want it equally in either case.

Reply:
I felt like this code doesn't apply anymore once the bp == workspace rule is in place. It seems to be guarding against the possibility of duplicate workspace names within a billing project but that should no longer be a problem since every billing project at this point should be unique once its taken off of the buffer.

If the retry was on `Exception` I might have kept it around but its only handling the conflict case

Reply:
Good point - I overlooked that this was just for conflict handling. Makes sense to fork.

Thinking more - there's actually an argument here that we use a constant for all firecloud workspace IDs. I think I prefer the way you have it now though.